Add first solution hw2-1

Nigma-Ks opened the pull request.

Nigma-Ks committed:
Add first solution hw2-1

Lozov-Petr commented on hw2_1/hw2-1/Program.cs:
```diff
@@ -0,0 +1,317 @@
+    {
+        static void Main(string[] args)
+        {
+            BorStringStorage storage = new BorStringStorage();
```
```suggestion
            BorStringStorage storage = new();
```

Lozov-Petr replied:
```suggestion
            var storage = new BorStringStorage();
```
Или так

Lozov-Petr commented on hw2_1/hw2-1/Program.cs:
```diff
@@ -0,0 +1,317 @@
+                {
+                    case "1":
+                        Console.Write("Enter string which you want to add: ");
+                        string stringToAdd = Console.ReadLine();
```
Метод `ReadLine` возвращает `string?`, но используется это значение как `string`. 
И это может оказаться проблемой.
Например, при чтении входных данных, перенаправленных из файла, `Console.ReadLine` вернёт `null`, если из файла больше нечего читать.
Общее правило таково: если возвращаемой значение типа `Nullable<T>`, то и работать с ним надо, как с `Nullable` значением.

Lozov-Petr commented on hw2_1/hw2-1/Program.cs:
```diff
@@ -0,0 +1,317 @@
+                                  "4. Find out how many strings in storage starts with your entered string\n" +
+                                  "5. Find out if storage contains entered string\n" +
+                                  "6. Exit\n");
+                Console.Write("Write number 1, 2, 3, 4, 5 or 6: ");
```
Копипаста.
Если нужно какое-либо действие делать несколько раз (например, напечатать help), то его стоит вынести в отельный метод

Lozov-Petr commented on hw2_1/hw2-1/Program.cs:
```diff
@@ -0,0 +1,317 @@
+        {
+            get { return size; }
+            private set { size = value; }
+        }
```
```suggestion
        public int Size { get; private set; } = 0;
```
Так короче, нагляднее, и не нужно явно заводить приватное поле `size`.

Lozov-Petr commented on hw2_1/hw2-1/Program.cs:
```diff
@@ -0,0 +1,317 @@
+        public bool Add(string element)
+        {
+            bool isNewString;
+            isNewString = InsertString(_root, element); //not null because constructs with no null root
```
```suggestion
            bool isNewString = InsertString(_root, element); //not null because constructs with no null root

```

Lozov-Petr commented on hw2_1/hw2-1/Program.cs:
```diff
@@ -0,0 +1,317 @@
+
+        public BorStringStorage()
+        {
+            Vertex[] currArrayOfVertexes = new Vertex[256];
```
В C# `char` представляет символы Юникода UTF-16 и, соответственно имеет 16 разрядов. 
Поэтому 256 - это мало. Например, на русских символах, программа просто падает с исключением `IndexOutOfRangeException`.
Можно конечно, выделить массив размера `2^16` для каждой вершины. Но это неэффективно будет. Вместо массива лучше использовать `Dictionary`: не имеет ограничений по размеру и, что важнее, выделяет память по мере необходимости, а не сразу под все символы.

Lozov-Petr commented on hw2_1/hw2-1/Program.cs:
```diff
@@ -0,0 +1,317 @@
+            public bool isTerminal;
+            public Vertex[] arrayOfRelatedVertexes;
+
+            public Vertex(bool isTerminal, Vertex[] arrayOfRelatedVertexes)
```
Странно, что сюда сам массив передается. Представление данных в вершине - это особенность его реализации, так что его можно (и нужно) скрыть от внешнего класса.

Lozov-Petr commented on hw2_1/hw2-1/Program.cs:
```diff
@@ -0,0 +1,317 @@
+                int codeOfCurrentChar = (int)element[i];
+                if (currVertex.arrayOfRelatedVertexes[codeOfCurrentChar] == null)
+                {
+                    bool isTerminal = i == element.Length - 1;
```
Общие данные для обеих веток условного оператора (в else ветке это же значение вычисляется в строке 214.
Можно вынести за пределы условного оператора.

Lozov-Petr reviewed (commented):
3.5

Nigma-Ks committed:
1. string -> string?
2. отдельный метод для возможных команд
3. переписала size
4. массивы -> словари
5. в конструктор передается только isTerminal

Nigma-Ks committed:
убрала size

Nigma-Ks committed:
Исправила ошибку добавления первой буквы, убрала лишнии присваивания Size
